diff --git a/src/rdcrc32.h b/src/rdcrc32.h
index 676cd7d2..082a81c7 100644
--- a/src/rdcrc32.h
+++ b/src/rdcrc32.h
@@ -76,7 +76,7 @@ extern "C" {
 typedef uint32_t rd_crc32_t;
 
 #if !WITH_ZLIB
-extern const rd_crc32_t crc_table[256];
+extern const rd_crc32_t rd_crc_table[256];
 #endif
 
 
@@ -131,7 +131,7 @@ static RD_INLINE RD_UNUSED rd_crc32_t rd_crc32_update(rd_crc32_t crc,
 
         while (data_len--) {
                 tbl_idx = (crc ^ *data) & 0xff;
-                crc     = (crc_table[tbl_idx] ^ (crc >> 8)) & 0xffffffff;
+                crc     = (rd_crc_table[tbl_idx] ^ (crc >> 8)) & 0xffffffff;
 
                 data++;
         }
