diff --git a/src/rdcrc32.c b/src/rdcrc32.c
index f7a68855..a9b99cf6 100644
--- a/src/rdcrc32.c
+++ b/src/rdcrc32.c
@@ -49,7 +49,7 @@
 /**
  * Static table used for the table_driven implementation.
  *****************************************************************************/
-const rd_crc32_t crc_table[256] = {
+const rd_crc32_t rd_crc_table[256] = {
     0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
     0xe963a535, 0x9e6495a3, 0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
     0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2,
