diff --git a/src/rdkafka_op.c b/src/rdkafka_op.c
index 34e9e3fd..55a15e14 100644
--- a/src/rdkafka_op.c
+++ b/src/rdkafka_op.c
@@ -739,6 +739,10 @@ rd_kafka_op_res_t
 rd_kafka_op_call(rd_kafka_t *rk, rd_kafka_q_t *rkq, rd_kafka_op_t *rko) {
         rd_kafka_op_res_t res;
         rd_assert(rko->rko_op_cb);
+/*madhu RD_TLS*/
+#ifdef __MVS__
+        GET_KEY_VAL(rd_kafka_yield_thread_key,rd_kafka_yield_thread_key_s,rd_kafka_yield_thread_p,sizeof(int))
+#endif
         res = rko->rko_op_cb(rk, rkq, rko);
         if (unlikely(res == RD_KAFKA_OP_RES_YIELD || rd_kafka_yield_thread))
                 return RD_KAFKA_OP_RES_YIELD;
