diff --git a/src/rdkafka_conf.c b/src/rdkafka_conf.c
index 39438e7..5fe921c 100644
--- a/src/rdkafka_conf.c
+++ b/src/rdkafka_conf.c
@@ -60,6 +60,24 @@
 #include <curl/curl.h>
 #endif
 
+#define ALLOWED_CURL_OPTS_COUNT 12
+
+static const char *allowed_curl_opts[ALLOWED_CURL_OPTS_COUNT] = {
+    "CURLOPT_SSLCERT",
+    "CURLOPT_SSLCERTPASSWD",
+    "CURLOPT_SSLKEYPASSWD",
+    "CURLOPT_SSLVERSION",
+    "CURLOPT_VERBOSE",
+    "CURLOPT_SSL_VERIFYPEER",
+    "CURLOPT_CAINFO",
+    "CURLOPT_SSL_VERIFYHOST",
+    "CURLOPT_SSL_CIPHER_LIST",
+    "CURLOPT_SSLCERTTYPE",
+    "CURLOPT_SSLKEY",
+    "CURLOPT_SSLKEYTYPE"
+};
+
+
 struct rd_kafka_property {
         rd_kafka_conf_scope_t scope;
         const char *name;
@@ -330,6 +348,67 @@ rd_kafka_conf_validate_broker_version(const struct rd_kafka_property *prop,
         return rd_kafka_get_legacy_ApiVersions(val, &apis, &api_cnt, NULL);
 }
 
+/* Function to check if a value is in the allowed_curl_opts array */
+static int is_valid_curl_opt(const char *opt) {
+    for (int i = 0; i < ALLOWED_CURL_OPTS_COUNT; ++i) {
+        if (strcmp(opt, allowed_curl_opts[i]) == 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static RD_UNUSED int
+rd_kafka_conf_validate_curl_opts(const struct rd_kafka_property *prop,
+                              const char *val,
+                              int ival) {
+    char *opts, *token, *saveptr;
+    int valid = 1; 
+
+    if (val == NULL || *val == '\0') {
+        return 0; 
+    }
+
+    opts = strdup(val);
+    if (opts == NULL) {
+        return 0; 
+    }
+
+    token = strtok_r(opts, ",", &saveptr);
+    while (token) {
+        char *equal_sign = strchr(token, '=');
+        if (equal_sign == NULL) {
+            valid = 0; 
+            break;
+        }
+
+        *equal_sign = '\0'; 
+        char *opt = token;
+        char *value = equal_sign + 1;
+
+        if (!is_valid_curl_opt(opt)) {
+            valid = 0; 
+            break;
+        }
+
+        // Validate the value part
+        for (char *p = value; *p; p++) {
+            if (*p == ',' || isspace((unsigned char)*p)) {
+                valid = 0; 
+                break;
+            }
+        }
+
+        if (!valid) break;
+
+        token = strtok_r(NULL, ",", &saveptr);
+    }
+
+    free(opts);
+    return valid;
+}
+
+
 /**
  * @brief Validate that string is a single item, without delimters (, space).
  */
@@ -1092,13 +1171,25 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
      _RK(sasl.oauthbearer.method),
      "Set to \"default\" or \"oidc\" to control which login method "
      "to be used. If set to \"oidc\", the following properties must also be "
-     "be specified: "
+     "specified: "
      "`sasl.oauthbearer.client.id`, `sasl.oauthbearer.client.secret`, "
      "and `sasl.oauthbearer.token.endpoint.url`.",
      .vdef = RD_KAFKA_SASL_OAUTHBEARER_METHOD_DEFAULT,
      .s2i  = {{RD_KAFKA_SASL_OAUTHBEARER_METHOD_DEFAULT, "default"},
               {RD_KAFKA_SASL_OAUTHBEARER_METHOD_OIDC, "oidc"}},
      _UNSUPPORTED_OIDC},
+    {_RK_GLOBAL, "sasl.oauthbearer.user.name", _RK_C_STR,
+     _RK(sasl.oauthbearer.user_name),
+     "The resource owner's name."
+     "Only used when `sasl.oauthbearer.method` is set to \"oidc\" and"
+     "sasl.oauthbearer.grant.type is set to \"password\".",
+     _UNSUPPORTED_OIDC},
+    {_RK_GLOBAL, "sasl.oauthbearer.user.password", _RK_C_STR,
+     _RK(sasl.oauthbearer.user_password),
+     "The resource owner's password."
+     "Only used when `sasl.oauthbearer.method` is set to \"oidc\" and"
+     "sasl.oauthbearer.grant.type is set to \"password\".",
+     _UNSUPPORTED_OIDC},
     {_RK_GLOBAL, "sasl.oauthbearer.client.id", _RK_C_STR,
      _RK(sasl.oauthbearer.client_id),
      "Public identifier for the application. "
@@ -1153,11 +1244,28 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
         "provider.",
         _UNSUPPORTED_OIDC,
         .vdef = RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS,
-        .s2i  = {{RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS,
+        .s2i  = {{RD_KAFKA_SASL_OAUTHBEARER_CLIENT_CREDENTIALS,
+                  "client_credentials"},
+                 {RD_KAFKA_SASL_OAUTHBEARER_PASSWORD, "password"},
+                 {RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS,
                   "client_credentials"},
                  {RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_JWT_BEARER,
                   "urn:ietf:params:oauth:grant-type:jwt-bearer"}},
     },
+    {_RK_GLOBAL, "sasl.oauthbearer.grant_type", _RK_C_ALIAS,
+     .sdef = "sasl.oauthbearer.grant.type"},
+    {_RK_GLOBAL, "sasl.oauthbearer.curl.opts", _RK_C_STR,
+     _RK(sasl.oauthbearer.curl_opts),
+     "Comma separated list of curl opts. "
+     "Only these are curl options are supported: CURLOPT_SSLCERT, CURLOPT_SSLCERTPASSWD,"
+     " CURLOPT_SSLKEYPASSWD, CURLOPT_SSL_VERIFYPEER,"
+     " CURLOPT_CAINFO, CURLOPT_SSL_VERIFYHOST, CURLOPT_SSL_CIPHER_LIST, CURLOPT_SSLCERTTYPE,"
+     " CURLOPT_SSLKEY, CURLOPT_SSLKEYTYPE."
+     "Only used when `sasl.oauthbearer.method` is set to \"oidc\".",
+     .sdef = "",
+     .validate = rd_kafka_conf_validate_curl_opts},
+    {_RK_GLOBAL, "sasl.oauthbearer.principal.name", _RK_C_ALIAS,
+     .sdef = "sasl.oauthbearer.sub.claim.name"},
     {_RK_GLOBAL, "sasl.oauthbearer.assertion.algorithm", _RK_C_S2I,
      _RK(sasl.oauthbearer.assertion.algorithm),
      "Algorithm the client should use to sign the assertion sent "
@@ -2496,7 +2604,6 @@ static int rd_kafka_anyconf_set(int scope,
         char estmp[1];
         const struct rd_kafka_property *prop;
         rd_kafka_conf_res_t res;
-
         if (!errstr) {
                 errstr      = estmp;
                 errstr_size = 0;
@@ -3998,6 +4105,22 @@ char *rd_kafka_conf_kv_get(const char *config,
 const char *
 rd_kafka_conf_finalize_oauthbearer_oidc_grant_type(rd_kafka_conf_t *conf) {
         switch (conf->sasl.oauthbearer.grant_type) {
+        case RD_KAFKA_SASL_OAUTHBEARER_PASSWORD:
+                if (!conf->sasl.oauthbearer.client_id)
+                        return "`sasl.oauthbearer.client.id` is mandatory when "
+                               "`sasl.oauthbearer.method=oidc` and "
+                               "`sasl.oauthbearer.grant.type=password` are set";
+
+                if (!conf->sasl.oauthbearer.user_name)
+                        return "`sasl.oauthbearer.user.name` is mandatory when "
+                               "`sasl.oauthbearer.method=oidc` and "
+                               "`sasl.oauthbearer.grant.type=password` are set";
+
+                if (!conf->sasl.oauthbearer.user_password)
+                        return "`sasl.oauthbearer.user.password` is mandatory "
+                               "when `sasl.oauthbearer.method=oidc` and "
+                               "`sasl.oauthbearer.grant.type=password` are set";
+                break;
         case RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS:
                 if (!conf->sasl.oauthbearer.client_id)
                         return "`sasl.oauthbearer.client.id` is "
@@ -4924,5 +5047,5 @@ int unittest_conf(void) {
 
         RD_UT_PASS();
 }
-
 /**@}*/
+
