diff --git a/src/rdkafka_queue.h b/src/rdkafka_queue.h
index eb329d1c..a049202b 100644
--- a/src/rdkafka_queue.h
+++ b/src/rdkafka_queue.h
@@ -905,7 +905,14 @@ rd_kafka_queue_t *rd_kafka_queue_new0(rd_kafka_t *rk, rd_kafka_q_t *rkq);
 
 void rd_kafka_q_dump(FILE *fp, rd_kafka_q_t *rkq);
 
+#ifdef __MVS__
+extern int *rd_kafka_yield_thread_p;
+extern pthread_key_t rd_kafka_yield_thread_key;
+extern int rd_kafka_yield_thread_key_s;
+#define rd_kafka_yield_thread (*rd_kafka_yield_thread_p)
+#else
 extern int RD_TLS rd_kafka_yield_thread;
+#endif
 
 
 
