diff --git a/src/rdkafka_request.c b/src/rdkafka_request.c
index 5ac7e0f1..08f67b7a 100644
--- a/src/rdkafka_request.c
+++ b/src/rdkafka_request.c
@@ -62,9 +62,20 @@ static const char *rd_kafka_actions_descs[] = {
 };
 
 const char *rd_kafka_actions2str(int actions) {
+#ifdef __MVS__
+	static tls_t tls_actstr;
+	static char *actstr = NULL;
+	size_t size_of_actstr = 128;
+	tls_actstr.sz = sizeof(char)*size_of_actstr;
+	actstr = __tlsValue(&tls_actstr);
+    rd_flags2str(actstr, sizeof(actstr), rd_kafka_actions_descs,
+                            actions);
+	return actstr;
+#else
         static RD_TLS char actstr[128];
         return rd_flags2str(actstr, sizeof(actstr), rd_kafka_actions_descs,
                             actions);
+#endif
 }
 
 
