diff --git a/src/rdkafka_ssl.c b/src/rdkafka_ssl.c
index 85f745cb..52d39d3e 100644
--- a/src/rdkafka_ssl.c
+++ b/src/rdkafka_ssl.c
@@ -1772,6 +1772,8 @@ static RD_UNUSED unsigned long rd_kafka_transport_ssl_threadid_cb(void) {
          * and thread id, which means we can't use the
          * thrd_current() API that returns the handle. */
         return (unsigned long)GetCurrentThreadId();
+#elif __MVS__
+        return (unsigned long)(intptr_t)thrd_current().__;
 #else
         return (unsigned long)(intptr_t)thrd_current();
 #endif
