diff --git a/src/rdkafka_conf.h b/src/rdkafka_conf.h
index f15d428..d12d965 100644
--- a/src/rdkafka_conf.h
+++ b/src/rdkafka_conf.h
@@ -155,8 +155,11 @@ typedef enum {
 } rd_kafka_oauthbearer_method_t;
 
 typedef enum {
-        RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS,
-        RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_JWT_BEARER,
+        RD_KAFKA_SASL_OAUTHBEARER_CLIENT_CREDENTIALS,
+        RD_KAFKA_SASL_OAUTHBEARER_PASSWORD,
+        RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_CLIENT_CREDENTIALS =
+            RD_KAFKA_SASL_OAUTHBEARER_CLIENT_CREDENTIALS,
+        RD_KAFKA_SASL_OAUTHBEARER_GRANT_TYPE_JWT_BEARER = 2,
 } rd_kafka_oauthbearer_grant_type_t;
 
 typedef enum {
@@ -382,6 +385,9 @@ struct rd_kafka_conf_s {
                         void (*token_refresh_cb)(rd_kafka_t *rk,
                                                  const char *oauthbearer_config,
                                                  void *opaque);
+                        char *user_name;
+                        char *user_password;
+                        char *curl_opts;
                 } oauthbearer;
         } sasl;
 
