diff --git a/src/rdposix.h b/src/rdposix.h
index 7b237682..c7804e4e 100644
--- a/src/rdposix.h
+++ b/src/rdposix.h
@@ -60,12 +60,15 @@
 #define RD_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
 #define RD_NORETURN           __attribute__((noreturn))
 #define RD_IS_CONSTANT(p)     __builtin_constant_p((p))
+#ifdef _ZOS
+#define RD_TLS  
+#else
 #define RD_TLS                __thread
-
+#endif
 /**
  * Allocation
  */
-#if !defined(__FreeBSD__) && !defined(__OpenBSD__)
+#if !defined(__FreeBSD__) && !defined(__OpenBSD__) && !defined(__MVS__)
 /* alloca(3) is in stdlib on FreeBSD */
 #include <alloca.h>
 #endif
@@ -136,7 +139,9 @@ static RD_INLINE RD_UNUSED const char *rd_strerror(int err) {
 /**
  * Misc
  */
-
+#ifdef _ZOS
+#define rd_usleep(usec, terminate) sleep((usec) / 1000)
+#else
 /**
  * Microsecond sleep.
  * Will retry on signal interrupt unless *terminate is true.
@@ -149,7 +154,7 @@ static RD_INLINE RD_UNUSED void rd_usleep(int usec, rd_atomic32_t *terminate) {
                (errno == EINTR && (!terminate || !rd_atomic32_get(terminate))))
                 ;
 }
-
+#endif
 
 
 #define rd_gettimeofday(tv, tz) gettimeofday(tv, tz)
