diff --git a/src/rdrand.c b/src/rdrand.c
index e36d7938..0d3c3b02 100644
--- a/src/rdrand.c
+++ b/src/rdrand.c
@@ -41,7 +41,11 @@ int rd_jitter(int low, int high) {
                 struct timeval tv;
                 rd_gettimeofday(&tv, NULL);
                 seed = (unsigned int)(tv.tv_usec / 1000);
+#ifdef _ZOS
+                seed ^= (unsigned int)(intptr_t)(thrd_current().__);
+#else
                 seed ^= (unsigned int)(intptr_t)thrd_current();
+#endif
         }
 
         rand_num = rand_r(&seed);
