diff --git a/src/rdrand.c b/src/rdrand.c
index bdab0029..2fca7fa9 100644
--- a/src/rdrand.c
+++ b/src/rdrand.c
@@ -30,18 +30,34 @@
 #include "rdrand.h"
 #include "rdtime.h"
 #include "tinycthread.h"
+#ifdef __MVS__
+#include <zos-tls.h>
+#endif
 
 int rd_jitter(int low, int high) {
         int rand_num;
 #if HAVE_RAND_R
+#ifdef __MVS__
+	static pthread_key_t seed_key;
+	static int seed_key_s = 0;
+        unsigned int *seed_p = NULL;
+	GET_KEY_VAL(seed_key,seed_key_s,seed_p,sizeof(unsigned int))
+#define seed (*seed_p)
+#else
         static RD_TLS unsigned int seed = 0;
+#endif
 
         /* Initial seed with time+thread id */
         if (unlikely(seed == 0)) {
                 struct timeval tv;
                 rd_gettimeofday(&tv, NULL);
                 seed = (unsigned int)(tv.tv_usec / 1000);
+#ifdef __MVS__
+                seed ^= (unsigned int)(intptr_t)(thrd_current().__);
+SET_KEY_VAL(seed_key,seed_p)
+#else
                 seed ^= (unsigned int)(intptr_t)thrd_current();
+#endif
         }
 
         rand_num = rand_r(&seed);
