diff --git a/src/tinycthread.c b/src/tinycthread.c
index b0ec8e95..4f351096 100644
--- a/src/tinycthread.c
+++ b/src/tinycthread.c
@@ -608,13 +608,21 @@ int thrd_create(thrd_t *thr, thrd_start_t func, void *arg)
           if((err = pthread_create(thr, NULL, _thrd_wrapper_function,
                                    (void *)ti)) != 0) {
                   errno = err;
+#ifdef __MVS__
+                  thr->__ = 0;
+#else
                   *thr = 0;
+#endif
           }
   }
 #endif
 
   /* Did we fail to create the thread? */
+#ifdef __MVS__
+  if(!(thr->__))
+#else
   if(!*thr)
+#endif
   {
     rd_free(ti);
     return thrd_error;
