diff --git a/src/tinycthread.h b/src/tinycthread.h
index 6bc39fe0..d3771a4f 100644
--- a/src/tinycthread.h
+++ b/src/tinycthread.h
@@ -482,9 +482,9 @@ int tss_set(tss_t key, void *val);
  * @param func Callback to invoke.
  */
 #if defined(_TTHREAD_WIN32_)
-  void call_once(once_flag *flag, void (*func)(void));
+  void rd_call_once(once_flag *flag, void (*func)(void));
 #else
-  #define call_once(flag,func) pthread_once(flag,func)
+  #define rd_call_once(flag,func) pthread_once(flag,func)
 #endif
 
 
