diff --git a/src/tinycthread_extra.c b/src/tinycthread_extra.c
index 1c62cdc..c2955a8 100644
--- a/src/tinycthread_extra.c
+++ b/src/tinycthread_extra.c
@@ -54,6 +54,8 @@ int thrd_setname(const char *name) {
 int thrd_is_current(thrd_t thr) {
 #if defined(_TTHREAD_WIN32_)
         return GetThreadId(thr) == GetCurrentThreadId();
+#elif defined (__MVS__)
+        return pthread_equal(pthread_self(), thr);
 #else
         return (pthread_self() == thr);
 #endif
@@ -65,6 +67,8 @@ unsigned long thrd_current_id(void) {
          * and thread id, which means we can't use the
          * thrd_current() API that returns the handle. */
         return (unsigned long)GetCurrentThreadId();
+#elif defined(__MVS__)
+        return (unsigned long)(intptr_t)thrd_current().__;
 #else
         return (unsigned long)(intptr_t)thrd_current();
 #endif
