diff --git a/examples/describe_topics.c b/examples/describe_topics.c
index 5b7425ef..bdc53f33 100644
--- a/examples/describe_topics.c
+++ b/examples/describe_topics.c
@@ -36,11 +36,13 @@
 #include <stdlib.h>
 #include <stdarg.h>
 
+#ifndef __MVS__
 #ifdef _WIN32
 #include "../win32/wingetopt.h"
 #else
 #include <getopt.h>
 #endif
+#endif
 
 
 /* Typical include path would be <librdkafka/rdkafka.h>, but this program
