diff --git a/examples/rdkafka_complex_consumer_example.c b/examples/rdkafka_complex_consumer_example.c
index ac56e659..d7f95cb7 100644
--- a/examples/rdkafka_complex_consumer_example.c
+++ b/examples/rdkafka_complex_consumer_example.c
@@ -40,7 +40,9 @@
 #include <syslog.h>
 #include <sys/time.h>
 #include <errno.h>
+#ifndef __MVS__
 #include <getopt.h>
+#endif
 
 /* Typical include path would be <librdkafka/rdkafka.h>, but this program
  * is builtin from within the librdkafka source tree and thus differs. */
