diff --git a/tests/0017-compression.c b/tests/0017-compression.c
index 8cb295f..f45a647 100644
--- a/tests/0017-compression.c
+++ b/tests/0017-compression.c
@@ -48,9 +48,11 @@ int main_0017_compression(int argc, char **argv) {
 #if WITH_ZLIB
                                              "gzip",
 #endif
+#ifndef __MVS__
 #if WITH_SNAPPY
                                              "snappy",
 #endif
+#endif
 #if WITH_ZSTD
                                              "zstd",
 #endif
