diff --git a/tests/0146-metadata_mock.c b/tests/0146-metadata_mock.c
index c0f1d7b1..9f9b0d73 100644
--- a/tests/0146-metadata_mock.c
+++ b/tests/0146-metadata_mock.c
@@ -197,7 +197,7 @@ static void do_test_stale_metadata_doesnt_migrate_partition(void) {
             RD_KAFKA_RESP_ERR_KAFKA_STORAGE_ERROR, 1000);
 
         /* Wait partition migrates to broker 2 */
-        rd_usleep(100 * 1000, 0);
+        rd_usleep(1000 * 1000, 0);
 
         /* Ask to return stale metadata while calling OffsetForLeaderEpoch */
         rd_kafka_mock_start_request_tracking(mcluster);
